### DISCUSSION TITLE
Netflix DGS Gateway Spec and References for RateLimiting/QueryComplexity/DepthLimit

### DISCUSSION THREAD

**Discussion by sbilello** [Netflix/dgs-framework#429] [General]
Hi!

1) Did you follow this specification https://www.apollographql.com/docs/federation/federation-spec/ to build the Netflix DGS Gateway? Maybe shall we consider building an open-source one in Java/Kotlin? 
I didn't find a Java Gateway in the market available so far.

2) In the case of Federation, the query entry point is the `apollo/gateway` so I would end up using the javascript libraries reported in this article https://moonhighway.com/securing-your-graphql-server to cover RateLimiting, QueryComplexity, and DepthLimit at the `apollo/gateway` level. 
Should we cover these aspects in some Netflix DGS modules for GraphQL setup without federation?

For the interested reader: I found that it is pretty easy to cover the query complexity via https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/analysis/MaxQueryComplexityInstrumentation.java.

**Comment by paulbakker** [comment]
1. Yes, our gateway is built using the federation spec, and have been working closely with Apollo to evolve the spec further. Unfortunately, there are currently no plans to open source the gateway.

2. This has come up before. Because most of our high-traffic DGSes are behind a gateway, there isn't much of a need for us at Netflix, but this could definitely be useful for others. I don't think we'll prioritize working on this any time soon, but would be happy to take pull requests.
